code/modules/reagents/chemistry/machinery/chem_heater.dm

```diff
@@ -83,17 +100,42 @@
 		updateUsrDialog()
 		update_icon()
 		return
+
+	if(beaker)
+		if(istype(I, /obj/item/reagent_containers/dropper))
+			var/obj/item/reagent_containers/dropper/D = I
+			D.afterattack(beaker, user, 1)
+			return
+		if(istype(I, /obj/item/reagent_containers/syringe))
+			var/obj/item/reagent_containers/syringe/S = I
+			S.afterattack(beaker, user, 1)
+			return
+	
 	return ..()
 
 /obj/machinery/chem_heater/on_deconstruction()
 	replace_beaker()
 	return ..()
 
+/obj/machinery/chem_heater/proc/on_reaction_step(datum/reagents/holder, num_reactions)
+	for(var/ui_client in ui_client_list)
+		var/datum/tgui/ui = ui_client
+		ui.send_update()
+
+
 /obj/machinery/chem_heater/ui_interact(mob/user, datum/tgui/ui)
 	ui = SStgui.try_update_ui(user, src, ui)
 	if(!ui)
 		ui = new(user, src, "ChemHeater", name)
 		ui.open()
+		LAZYADD(ui_client_list, ui)
+
```

Comment:
There's no risk of these datums being destroyed while in the list, is there?

Reply:
I'm honestly not sure. I can only use the same UI so much on my own. How can I tell?

Reply:
Maybe closing the game with them open might leave them in the list..? I could add a check to remove them when the beaker is removed too.

Reply:
Actually no I can't. Hmm.